File: frontend/src/services/geocodingService.js
// frontend/src/services/geocodingService.js
// Uses OpenStreetMap Nominatim API directly (no backend required)

const NOMINATIM_BASE = 'https://nominatim.openstreetmap.org';

const geocodingService = {
  /**
   * Search for locations by name using Nominatim
   * Returns array of { lat, lon, display_name, place_id, type, ... }
   */
  searchLocation: async (query) => {
    try {
      const params = new URLSearchParams({
        q: query,
        format: 'json',
        addressdetails: '1',
        limit: '8',
        countrycodes: 'in', // Bias results to India — remove if global search needed
      });

      const response = await fetch(
        `${NOMINATIM_BASE}/search?${params.toString()}`,
        {
          headers: {
            // Nominatim requires a descriptive User-Agent
            'Accept-Language': 'en',
          },
        }
      );

      if (!response.ok) {
        throw new Error(`Nominatim error: ${response.status}`);
      }

      const data = await response.json();

      return {
        success: true,
        data: data.map((item) => ({
          lat: item.lat,
          lon: item.lon,
          display_name: item.display_name,
          place_id: item.place_id,
          type: item.type,
          importance: item.importance,
        })),
      };
    } catch (error) {
      console.error('Location search failed:', error);
      return { success: false, data: [], error: error.message };
    }
  },

  /**
   * Reverse geocode: coordinates → address string
   */
  getAddressFromCoords: async (lat, lng) => {
    try {
      const params = new URLSearchParams({
        lat: lat.toString(),
        lon: lng.toString(),
        format: 'json',
      });

      const response = await fetch(
        `${NOMINATIM_BASE}/reverse?${params.toString()}`
      );

      if (!response.ok) throw new Error(`Nominatim reverse error: ${response.status}`);

      const data = await response.json();
      return { success: true, data: data.display_name };
    } catch (error) {
      console.error('Reverse geocoding failed:', error);
      return { success: false, data: null };
    }
  },

  /**
   * Legacy compatibility shim — calls searchLocation internally
   */
  getCoordinates: async (location) => {
    const result = await geocodingService.searchLocation(location);
    if (result.success && result.data.length > 0) {
      const first = result.data[0];
      return {
        success: true,
        data: { lat: parseFloat(first.lat), lng: parseFloat(first.lon), name: first.display_name },
      };
    }
    return { success: false, data: null };
  },
};

export default geocodingService;

File: neurofleetx-backend/src/main/java/com/neurofleetx/service/GeocodingService.java
package com.neurofleetx.service;

import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class GeocodingService {

    private final RestTemplate restTemplate = new RestTemplate();

    public Map<String, Object> searchLocation(String locationName) {
        try {
            // Using Nominatim (OpenStreetMap) API - Free, no API key needed
            String url = UriComponentsBuilder
                    .fromHttpUrl("https://nominatim.openstreetmap.org/search")
                    .queryParam("q", locationName)
                    .queryParam("format", "json")
                    .queryParam("limit", "5")
                    .toUriString();

            List<Map<String, Object>> response = restTemplate.getForObject(url, List.class);

            Map<String, Object> result = new HashMap<>();
            result.put("success", true);
            result.put("data", response);

            return result;

        } catch (Exception e) {
            Map<String, Object> error = new HashMap<>();
            error.put("success", false);
            error.put("message", "Failed to search location: " + e.getMessage());
            return error;
        }
    }

    public Map<String, Double> getCoordinates(String locationName) {
        try {
            String url = UriComponentsBuilder
                    .fromHttpUrl("https://nominatim.openstreetmap.org/search")
                    .queryParam("q", locationName)
                    .queryParam("format", "json")
                    .queryParam("limit", "1")
                    .toUriString();

            List<Map<String, Object>> response = restTemplate.getForObject(url, List.class);

            if (response != null && !response.isEmpty()) {
                Map<String, Object> location = response.get(0);
                Map<String, Double> coordinates = new HashMap<>();
                coordinates.put("lat", Double.parseDouble(location.get("lat").toString()));
                coordinates.put("lng", Double.parseDouble(location.get("lon").toString()));
                return coordinates;
            }

            return null;

        } catch (Exception e) {
            return null;
        }
    }
}